### DISCUSSION TITLE
Help Sort Order

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3909] [Development]
## Current State

NeoMutt has a Help Page built in.
Hit <kbd>?</kbd> on most screens an it'll bring up a page listing, for _that_ screen:

- All the current bindings
- All the current macros
- All the other functions that you _could_ use

It's very useful, but its layout is very confusing.
Go have a look now.

The order of the bindings/macros is:

- <kbd>\<Ctrl-A\></kbd>, <kbd>\<Ctrl-B\></kbd>, ...
- <kbd>\<Esc\>a</kbd> (aka <kbd>\<Alt-A\></kbd>), ...
- _Some_ symbols
- <kbd>A</kbd>, <kbd>B</kbd>, <kbd>C</kbd>, ..., <kbd>X</kbd>, <kbd>Y</kbd>, <kbd>Z</kbd>
- Some _more_ symbols
- <kbd>a</kbd>, <kbd>b</kbd>, <kbd>c</kbd>, ..., <kbd>x</kbd>, <kbd>y</kbd>, <kbd>z</kbd>
- Yet more symbols
- Function keys, like <kbd>\<Down\></kbd>, <kbd>\<Up\></kbd>, <kbd>\<F3\></kbd>

Humans don't think like this - <kbd>\<Up\></kbd> doesn't come between <kbd>\<Down\></kbd> and <kbd>\<F3\></kbd>.

Finally, there's a separation between an **untitled** first section and the "**Generic bindings**" below.

## A New Order

The user wants a concise list of keybindings.

First, I suggest combining the **untitled** and "**Generic bindings**" into one block.
Then, I propose a slightly complicated sorting strategy.

For the first level of sorting, modifiers are ignored.

1. Alphabetic
   <kbd>a</kbd>, <kbd>A</kbd>, <kbd>b</kbd>, <kbd>B</kbd>, <kbd>c</kbd>, <kbd>C</kbd>, ...
2. Symbols sorted by ASCII value, e.g.
   <kbd>%</kbd>, <kbd>,</kbd>, <kbd>.</kbd>, <kbd>@</kbd>, ...
3. Function Keys sorted by their text name, e.g.
   <kbd>\<Down\></kbd>, <kbd>\<F3\></kbd>, <kbd>\<Up\></kbd>

Then, if two bindings have the same key, then the modifier is used (sorted by their text name)
- \[none\]
- <kbd>Alt-</kbd>
- <kbd>Ctrl-</kbd>
- <kbd>Shift-</kbd>

## Example Help Page

| Key                        | Function                    |
| :------------------------- | :-------------------------- |
| <kbd>a</kbd>               | `<current-bottom>`          |
| <kbd>\<Alt-a\></kbd>       | `<followup-message>`        |
| <kbd>A</kbd>               | `<current-top>`             |
| <kbd>\<Alt-A\></kbd>       | `<edit-raw-message>`        |
| <kbd>\<Ctrl-A\></kbd>      | `<current-middle>`          |
| <kbd>b</kbd>               | `<imap-logout-all>`         |
| <kbd>\<Alt-b\></kbd>       | `<modify-labels-then-hide>` |
| <kbd>B</kbd>               | `<modify-labels>`           |
| <kbd>\<Alt-B\></kbd>       | `<root-message>`            |
| <kbd>\<Ctrl-B\></kbd>      | `<next-unread-mailbox>`     |
| <kbd>c</kbd>               | `<reconstruct-thread>`      |
| <kbd>\<Alt-c\></kbd>       | `<sidebar-toggle-virtual>`  |
| <kbd>C</kbd>               | `<search-opposite>`         |
| <kbd>\<Alt-C\></kbd>       | `<imap-fetch-mail>`         |
| <kbd>\<Ctrl-C\></kbd>      | `<previous-unread>`         |
|                            |                             |
| <kbd>%</kbd>               | `<toggle-write>`            |
| <kbd>,</kbd>               | `<macro>`                   |
| <kbd>.</kbd>               | `<mailbox-list>`            |
| <kbd>@</kbd>               | `<display-address>`         |
| <kbd>\<Alt-@\></kbd>       | `<sync-mailbox>`            |
|                            |                             |
| <kbd>\<Down\></kbd>        | `<next-undeleted>`          |
| <kbd>\<Ctrl-Down\></kbd>   | `<next-undeleted>`          |
| <kbd>\<Shift-Down\></kbd>  | `<next-undeleted>`          |
| <kbd>\<F3\></kbd>          | `<macro>`                   |
| <kbd>\<F4\></kbd>          | `<macro>`                   |
| <kbd>\<Tab\></kbd>         | `<next-new-then-unread>`    |
| <kbd>\<Shift-Tab\></kbd>   | `<previous-undeleted>`      |

This would be followed by a list of unbound functions.

---

**Updates**:
- @mschilli87's idea: group a's then A's

**Comment by mschilli87** [comment]
Sounds like a good idea. One nitpick: I'd swap the priority of case and modifiers:
`a`, `A`, `<Alt-a>`, `<Alt-A>`, `<Ctrl-a>`, `<Ctrl-A>` -> `a`, `<Alt-a>`,`<Ctrl-a>`, `A`, `<Alt-A>`, `<Ctrl-A>`

My reasoning is two-fold:

1. (*subjective!*:) It feels much more intuitive (to me) to have all `a` bindings grouped together rather than have them interspersed with `A` bindings.
2. (*objective?*:) I consider `A` equivalent to `<Shift-a>`. Thus, my suggestion better aligns the `a`/`A` bindings to the 'function key' ones (like `<Down>`, `<Alt-Down>`, `<Ctrl-Down>`, `<Shift-Down>`).

**Reply by flatcap** [reply]
Yes, that's better.
Post updated.

**Comment by ossilator** [comment]
i think that the order is completely irrelevant in the first place, and you could randomize it just as well - the only reason for me to open this page is to find the binding for a function i rarely use. for that i immediately use `/`.

so for me an actual improvement would mean making the search more efficient:
- (incremental) filtering, so i don't have to scroll through the text looking for highlights
- possibly have some "hidden" keywords outside the function name and description. these may not match mutt terminology, but would be still naturally associated with the action (e.g., 'write' for 'compose', always alias 'mailbox' and 'folder', etc.).

**Reply by flatcap** [reply]
> i think that the order is completely irrelevant in the first place

hehe, anarchy!

> to find the binding for a function i rarely use

I actually find myself working the other way.
Someone asks for help with a function to do something and I know the key, but not the exact function name.

> search more efficient
> (incremental) filtering

Ah!  We may have the beginnings of this in #2397 - add search possibility to sidebar
I'd thought about using it for the Index and Address Book, but Help's another good use-case.

 > possibly have some "hidden" keywords outside the function name and description

That's a great idea.  Thanks!  Noted.

Another subtle variation would be to always check both the English text and the translation.

I'll start investigating...
